go.mod

```diff
@@ -49,32 +49,32 @@ require (
 	golang.org/x/mod v0.27.0
 	golang.org/x/text v0.28.0
 	gopkg.in/yaml.v3 v3.0.1
-	k8s.io/api v0.32.3
-	k8s.io/apimachinery v0.32.3
-	k8s.io/client-go v0.32.3
+	k8s.io/api v0.33.4
+	k8s.io/apimachinery v0.33.4
+	k8s.io/client-go v0.33.4
 	k8s.io/cluster-bootstrap v0.32.3
-	k8s.io/component-base v0.32.3
+	k8s.io/component-base v0.33.4
 	k8s.io/klog/v2 v2.130.1
 	k8s.io/kubectl v0.32.3
 	k8s.io/utils v0.0.0-20250321185631-1f6e0b77f77e
 	sigs.k8s.io/cloud-provider-azure v1.32.3
 	sigs.k8s.io/cluster-api v1.10.4
 	sigs.k8s.io/cluster-api/test v1.10.4
-	sigs.k8s.io/controller-runtime v0.20.4
+	sigs.k8s.io/controller-runtime v0.21.0
```

Comment:
FYI @mboersma in case the current version of ASO doesn't build against this newer controller-runtime in #5661, the changes in this PR will need to be incorporated there. If CI is green without bumping ASO there though, I don't think we need to try to do that.